plugin/pkg/scheduler/algorithm/priorities/metadata.go

```diff
@@ -57,43 +58,63 @@ func (pmf *PriorityMetadataFactory) PriorityMetadata(pod *v1.Pod, nodeNameToInfo
 	if pod == nil {
 		return nil
 	}
-	tolerationsPreferNoSchedule := getAllTolerationPreferNoSchedule(pod.Spec.Tolerations)
-	podSelectors := getSelectors(pod, pmf.serviceLister, pmf.controllerLister, pmf.replicaSetLister, pmf.statefulSetLister)
 	return &priorityMetadata{
-		nonZeroRequest: getNonZeroRequests(pod),
-		podTolerations: tolerationsPreferNoSchedule,
-		affinity:       pod.Spec.Affinity,
-		podSelectors:   podSelectors,
-		controllerRef:  priorityutil.GetControllerRef(pod),
+		nonZeroRequest:          getNonZeroRequests(pod),
+		podTolerations:          getAllTolerationPreferNoSchedule(pod.Spec.Tolerations),
+		affinity:                pod.Spec.Affinity,
+		podSelectors:            getSelectors(pod, pmf.serviceLister, pmf.controllerLister, pmf.replicaSetLister, pmf.statefulSetLister),
+		controllerRef:           priorityutil.GetControllerRef(pod),
+		podFirstServiceSelector: getFirstServiceSelector(pod, pmf.serviceLister),
 	}
 }
 
+// getFirstServiceSelector returns one selector of services the given pod.
+func getFirstServiceSelector(pod *v1.Pod, sl algorithm.ServiceLister) (firstServiceSelector labels.Selector) {
+	if sl != nil {
+		if services, err := sl.GetPodServices(pod); err == nil && len(services) > 0 {
+			return labels.SelectorFromSet(services[0].Spec.Selector)
+		}
+	}
+	return nil
+}
+
 // getSelectors returns selectors of services, RCs and RSs matching the given pod.
 func getSelectors(pod *v1.Pod, sl algorithm.ServiceLister, cl algorithm.ControllerLister, rsl algorithm.ReplicaSetLister, ssl algorithm.StatefulSetLister) []labels.Selector {
 	var selectors []labels.Selector
-	if services, err := sl.GetPodServices(pod); err == nil {
-		for _, service := range services {
-			selectors = append(selectors, labels.SelectorFromSet(service.Spec.Selector))
+
+	if sl != nil {
```

Comment:
Hmm - I'm against introducing those "ifs" here.
If those listers are not passed, this is a bug in code and should be fixed. With those ifs, we are risking not observing this.

Reply:
Actually, these "ifs" are also avoid of panic in test case "TestZoneSpreadPriority". In this test case rsl/ssl/cl are all set to nil expect service. if we set a specific non-useful rsl/ssl/cl to avoid this panic in this test case, looks like redundant is some degree.

Reply:
Let's not add production code to fix test issues. Change test in that case.

Reply:
And if the test would be the same as another one, let's remove it.

Reply:
I have added three local variables (sss/rcs/rss)  in TestZoneSpreadPriority to make sure the parameter not nil when construct metaDataProducer. and remove the  related "ifs".